rfcs/20181016-replicator.md

```diff
@@ -0,0 +1,1053 @@
+
+    Args:
+      value: The nested structure of `Tensor`s to gather across replicas.
+        The structure must be compatible with `tf.nest`.
+
+    Returns:
+       A `Tensor` nest with the `value`s from each replica. If a `Tensor` within
+       `value` has shape `dims`, the returned `Tensor` will have shape
+       `[num_replicas_in_sync] + `dims`.
+    """
+
+  def broadcast(self, value: T, source_replica_id: int) -> T:
+    """Broadcasts the given `Tensor` nest from the source to all replicas.
+
+    If `broadcast` is called in any replica, it must be called in all replicas.
+    The nested structure, `Tensor` shapes, and `source_replica_id` must be
+    identical in all replicas.
+
+    IMPORTANT: The ordering of communications must be identical in all replicas. See class docstring.
+
+    `value` will always be evaluated on all replicas.
+
+    Example with two replicas:
+      Replica 0:: `value`: {'a': 1, 'b': [40,  1]}, `source_replica_id`: 1
+      Replica 1:: `value`: {'a': 3, 'b': [ 2, 98]}, `source_replica_id`: 1
+
+      Replica 0:: result: {'a': 3, 'b': [ 2, 98]}
+      Replica 1:: result: {'a': 3, 'b': [ 2, 98]}
+
+    Args:
+      value: The nested structure of `Tensor`s to broadcast.
+        The structure must be compatible with `tf.nest`.
+      source_replica_id: The ID of the replica from which to broadcast.
+
+    Returns:
+      A `Tensor` nest with the `value` from the source replica.
+    """
+
+  def merge_call(merge_fn, *args, **kwargs):
+    """Merge args across replicas and run `merge_fn` in a cross-replica context.
+
+    IMPORTANT: The ordering of `merge_call`s must be identical in all replicas. See class docstring.
+    """
+```
+
+#### AggregationType enum
+
+```python
+class AggregationType(enum.Enum):
+  SUM = 1
+  MEAN = 2
+  ONLY_FIRST_REPLICA = 3
+  MIN = 4
+  MAX = 5
+```
+
+
+#### InputReplicationMode enum
+
+```python
+class InputReplicationMode(enum.Enum):
+  SINGLE = 0
+  PER_WORKER = 1
+  PER_REPLICA = 2
+```
+
+##### SINGLE
+
+The input function will be called once, typically on the first worker. The
+entire input pipeline ops then live on that worker. All replicas (on that and
+other workers) will dequeue from a single `Dataset` created on that worker.
+
+This mode is not supported by between-graph `DistributionStrategy`
+implementations.
+
+##### PER_WORKER
+
+The input function will be called on each worker independently, creating as many
+input pipelines as number of workers. Replicas will dequeue from the local
+`Dataset` on their worker. DS doesn't manage any state sharing between
+such separate input pipelines.
+
+##### PER_REPLICA
+
+The input function will be called for every replica, on the corresponding
+worker, thus creating as many input pipelines as number of replicas. Each
+replica will dequeue from its own `Dataset`.
+
+#### InputContext class
+
+An instance of `InputContext` is passed to the input function.
+
+```python
+class InputContext(object):
```

Comment:
Thanks for the review @wangsiyu . 
For Keras, we have had some discussions around whether we should allow users passing an input function. From the feedback we've got so far, it seems that providing a dataset or numpy array directly, and letting DS figure out how to split/shard the input is the most user friendly approach for Keras APIs. So we are not planning to change Keras API to take an input function. We will instead work on efficient distribution of input from a single dataset. 

For Estimator, this is under discussion right now, since it already supports input function. As you mentioned, it doesn't currently take this input context, but it would be reasonable to modify that to allow it for users who do wish to modify their input pipelines. Most likely we will add this support. 
However, note that this would require users to modify their existing input code. Until now, our philosophy has been to allow keras/estimator users to be able to use their existing code as much as possible when adding DS. So we are trying to figure out what's the best user experience for users who don't wish to modify their existing input pipelines.